sql/core/src/main/scala/org/apache/spark/sql/sources/DataSourceStrategy.scala

```diff
@@ -176,6 +176,9 @@ private[sql] object DataSourceStrategy extends Strategy {
       case expressions.Contains(a: Attribute, Literal(v: String, StringType)) =>
         Some(sources.StringContains(a.name, v))
 
+      case p: expressions.Predicate =>
```

Comment:
I think I agree.  It is reasonable to have predicates that can't be pushed down for various reasons.  Another issue is this check could often be a runtime error instead of some static compile time check.